docs/CIP_6_Pipelines_Registry.md

```diff
@@ -0,0 +1,309 @@
+collection.add(data=["small_chunk_1", "small_chunk_2"], ids=[0, 1])
+# Now add large documents with chunking
+collection.add(data=["large_document_1", "large_document_2"], ids=[2, 3], pipeline=DefaultEmbeddingFunctionPipelineStep(Chunk()))
+```
+
+You could also run the pipeline manually and then pass the output to the add step. For example:
+
+```python
+collection = client.create_collection("my_collection", pipeline=DefaultEmbeddingFunctionPipelineStep())
+# Add initial data
+collection.add(data=["small_chunk_1", "small_chunk_2"], ids=[0, 1])
+# Now add large documents with chunking but manually run the pipeline
+pipeline = DefaultEmbeddingFunctionPipelineStep(Chunk(["large_document_1", "large_document_2"]))
+data = pipeline.run()
+collection.add(data=data, ids=[2, 3])
+```
+
+#### Server-side registration of pipeline steps
+In order to register pipeline steps on the server, you will have to provide a python file that contains the pipeline steps, registered with the
+@pipeline_step decorator. For now, these are defined statically with each deploy of the server. In the future, we could allow for dynamic registration of pipeline steps.
+
+#### Future work: Output pipelines
+While not explicitly a part of this CIP, in the future, we could also add Output pipelines that are run on the output
+of a query. This would allow things like re-ranking, or filtering of the results. For example, if you wanted to re-rank the results of a query, you could do something like this:
+
+```python
+collection = client.create_collection("my_collection", pipeline=DefaultEmbeddingFunctionPipelineStep(), output_pipeline=ReRank())
+# Add initial data
+collection.add(data=["small_chunk_1", "small_chunk_2"], ids=[0, 1])
+# Query the collection
+collection.query(data=["query"], n_results=100) # ReRank will run before the results are returned
+```
+
+NOTE: Generically, one can view the existing behavior of chroma as a pipeline that runs metadata filtering followed by vector search. In the future, we could allow exposing this basic pipeline to the user as the default.
+
+This abstraction would also allow for the full RAG loop to be implemented in Chroma. For example, if you wanted to run a query, and then generate a response using a language model, you could do something like this:
+
+```python
+collection.query(data=["query"], n_results=100, output_pipeline=Generate(Composite(prompt)))
+```
+
+This pipeline first runs a partially initialized composite step on the results of the query, and then runs the generate step on the output of the composite step. This allows for the full RAG loop to be implemented in Chroma. The composite step combines the prompt and the retrieved documents, and the generate step generates the response.
+
+#### Future work: Parallel pipeline execution
+Not all pipeline stages need to executed sequentially. In the future, we can add support for parallel pipeline execution by changing how the pipeline is defined.
+
+## Storage Pipelines
```

Comment:
@HammadB, just a thought: Can this be, in theory, called storage adapters/interfaces?

An adapter, as in a wrapper on top of a specific storage medium?

Reply:
Sure